File: java/sys_nat/fan/sys/BoolArray.java
package fan.sys;

import fanx.main.Sys;
import fanx.main.Type;

public class BoolArray {
	private final long size;
	private final int[] words;

	BoolArray(long size) {
		this.size = size;
		this.words = new int[((int) size >> 0x5) + 1];
	}

	public static BoolArray make(long size) {
		BoolArray self = new BoolArray(size);
		return self;
	}

	public boolean get(long index) {
		int i = (int)index;
	    return (words[i>>0x5] & (1 << (i & 0x1F))) != 0;
	}

	public void set(long index, boolean v) {
		int i = (int)index;
	    int mask = 1 << (i & 0x1F);
	    if (v)
	      words[i>>0x5] |= mask;
	    else
	      words[i>>0x5] &= ~mask;
	}

	public long size() {
		return this.size;
	}

	public BoolArray realloc(long newSize) {
		if (size == newSize)
			return this;
		BoolArray na = BoolArray.make(newSize);
		int len = words.length > na.words.length ? na.words.length : words.length;
		System.arraycopy(words, 0, na.words, 0, len);
		return na;
	}

	public BoolArray copyFrom(BoolArray that, long thatOffset, long thisOffset, long length) {
		System.arraycopy(that.words, (int) thatOffset, words, (int) thisOffset, (int) length);
		return this;
	}

	public final Type typeof() {
		if (type == null)
			type = Sys.findType("sys::BoolArray");
		return type;
	}

	private static Type type;

	public BoolArray fill(boolean val, long times) {
		int t = (int) times;
		for (int i = 0; i < t; ++i) {
			set(i, val);
		}
		return this;
	}

}

File: src/sys/js/BoolArray.js


fan.sys.BoolArray = fan.sys.Obj.$extend(fan.sys.Obj);
fan.sys.BoolArray.prototype.$ctor = function() {}
fan.sys.BoolArray.prototype.$typeof = function() { return fan.sys.BoolArray.$type; }

fan.sys.BoolArray.make = function(size) {
	self = new fan.sys.BoolArray()
	self.m_array = new Int8Array(size);
	return self
}

fan.sys.BoolArray.prototype.get = function(pos) {
	return this.m_array[pos]
}

fan.sys.BoolArray.prototype.set = function(pos, val) {
	this.m_array[pos] = val
}

fan.sys.BoolArray.prototype.size = function() {
	return this.m_array.length;
}

fan.sys.BoolArray.prototype.realloc = function(newSize) {
	if (this.m_array.length == newSize) return this;

	if (newSize > this.m_array.length) {
		for (i = this.m_array.length; i<newSize; ++i) {
			this.m_array.push(null)
		}
		return this;
	}

	na = fan.sys.BoolArray.make(newSize, this.m_of);
	len = this.m_array.length > newSize ? newSize : this.m_array.length;
	for (var i = 0; i<len; ++i) {
		na.m_array[i] = this.m_array[i]
	}
	return na;
}

fan.sys.BoolArray.prototype.fill = function(val, times) {
	for (var i = 0; i < times; ++i) {
		this.m_array[i] = val;
	}
	return this;
}

fan.sys.BoolArray.prototype.copyFrom = function(that, thatOffset, thisOffset, length) {
	if (this === that) {
		this.m_array.copyWithin(thisOffset, thatOffset, thatOffset+length)
		return this;
	}

	for (var i = 0; i<length; ++i) {
		this.m_array[thisOffset + i] = that.m_array[i+thatOffset]
	}
	return this;
}

